### ISSUE TITLE
Malformed HTTP method causes exception in twisted.web.http server/Resource.render

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8102]
|<img alt="Javantea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Javantea reported|
|-|-|
|Trac ID|trac#8102|
|Type|defect|
|Created|2015-11-12 03:48:37Z|
|Branch|https://github.com/twisted/twisted/tree/http-server-invalid-method-8102-2|

An encoding issue occurs in twisted.python.compat.nativeString that affects the Twisted Web Server.

[twisted/web/resource.py#L243 Resource.render](twisted/web/resource.py#L243 Resource.render) is the code that calls nativeString.

[twisted/python/compat.py#L373 nativeString](twisted/python/compat.py#L373 nativeString) is the true source of the bug. It should not decode with ascii if it doesn't need to. The input vector is valid utf-8 which isn't valid python2 but is valid python3.

Stack trace:
```
Unhandled Error
Traceback (most recent call last):
  File "twisted/protocols/basic.py", line 571, in dataReceived
    why = self.lineReceived(line)
  File "twisted/web/http.py", line 1678, in lineReceived
    self.allContentReceived()
  File "twisted/web/http.py", line 1757, in allContentReceived
    req.requestReceived(command, path, version)
  File "twisted/web/http.py", line 823, in requestReceived
    self.process()
--- <exception caught here> ---
  File "twisted/web/server.py", line 183, in process
    self.render(resrc)
  File "twisted/web/server.py", line 234, in render
    body = resrc.render(self)
  File "twisted/web/resource.py", line 243, in render
    m = getattr(self, 'render_' + nativeString(request.method), None)
  File "twisted/python/compat.py", line 373, in nativeString
    s.decode("ascii")
exceptions.UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 1: ordinal not in range(128)
```

repro:
```python
from twisted.web import server, xmlrpc
from twisted.internet import reactor
reactor.listenTCP(4009, server.Site(xmlrpc.XMLRPC()))
reactor.run()
```

```
#!sh
echo -n $'G\xc2\xa9ET / HTTP/1.1\r\nHost: localhost\r\n\r\n' | nc localhost 4009
```

These types of exceptions that give verbose information to an attacker speed up recon against a target. Naturally the severity is low because it's just an exception, but it's important to close these bugs in libraries so that they don't affect large numbers of users.

Attachments:

* [8102.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d48/d48b60f110d901e200eeb1774e735c9a39521d42/7e87c5493be64bf67cf64597565f6bb8a5720173.diff) (1813 bytes) - added by pawelmhm on 2015-11-15 12:57:12Z - 
* [8102-updated.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d48/d48b60f110d901e200eeb1774e735c9a39521d42/4bcf2c1d50a4e79ba3bd482fee96e806878216ce.diff) (2046 bytes) - added by pawelmhm on 2015-11-18 12:43:51Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8102 8102
type__defect defect
reporter__Javantea Javantea
priority__normal normal
milestone__None None
branch__branches_http_server_invalid_method_8102_2 branches/http-server-invalid-method-8102-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1447300117770360 1447300117770360
changetime__1447959145311673 1447959145311673
version__None None
owner__pawelmhm pawelmhm

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_2"></a>|@adiroiban commented|
|-|-|

(In [46238]) Branching to http-server-invalid-method-8102.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @pawelmhm|
|-|-|

Many thanks for your patch.

Please note that the twisted/web/topfiles/8102.bugfix file needs a text describing the bugfix.
This text will be part of the final release notes.

Based on my understanding of RFC 2616, HTTP method can be US-ASCII or ISO-8859-1

So I think that we should accept latin1 encoding.

For non US-ASCII or non latin1 encoding, I think that we should reject the request much earlier. That is when we first parse the request line, and not when we try to process the request.

I think that the test should be updated to not only talk about non-ascii but also about non-latin1.

We might need a test to check that latin1 are accepted.

I assume that people will try to keep US-ASCII for their method names, so we might want to reject even valind latin1 methods.

I have applied your path and sent it to buildslaves.

What do you think?

Waiting for your feedback.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

From pawelmhm via the mailing list, due to spam-filter unfortunateness:

----------

Thanks for review adiroiban!

> I assume that people will try to keep US-ASCII for their method names, so we might want to reject even valind latin1 methods.

I agree with that part. All standard HTTP method names are ascii, so I dont see much reason to support non-ascii HTTP methods. They were not supported before this patch (they caused failures), if we fix that and allow latin1 we will allow characters that are outside ascii in method names, e.g. users will be able to have non-standard latin1 method name: 'GET£'. In my opinion it's better to only allow ascii and nothing else.

> For non US-ASCII or non latin1 encoding, I think that we should reject the request much earlier. That is when we first parse the request line, and not when we try to process the request

Yes I agree, so patch should go somewhere to twisted.web.http.HTTPChannel.linereceived when we first get method (line 1709) right? Also is it ok we raise 501? With current patch we raise 501 Not Suppoted but maybe 400 Bad Request would be better here?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

Well... from my side latin-1 HTTP methods should be supported so that we are RFC complaint.
I think that this is the whole point of a RFC :)

 twisted.web.http.Request.method is initialized as a text with ``` method = "(no method yet)" ``` but later handles as bytes.... and in the docstring, the type is not declared.

I think that method should be bytes as this is what most other Twisted developers want... even if I would like it to be text.

400 and raising it from lineReceived is better :)

---------

You might want to continue this conversation over the mailing list or on #twisted-dev

------

Looking forward for a new patch.

Many thanks for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)<a name="note_6"></a>|@pawelmhm commented|
|-|-|

> Well... from my side latin-1 HTTP methods should be supported so that we are RFC complaint

Hmm you are right, we should be standard-compliant, but which part of standard allows non-ascii characters in HTTP method? Section 5 of RFC2616 which talks about request line and method says nothing about latin1 and all methods listed there have names that are ascii-US.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

OK. I could not find this in the RFC 2616 ... so lets go for ASCII only. Thanks and sorry for the confusion!

latin-1 is only for the entity-body



https://tools.ietf.org/html/rfc7230#section-3.2.4

```
   Historically, HTTP has allowed field content with text in the
   ISO-8859-1 charset [ISO-8859-1], supporting other charsets only
   through use of [RFC2047] encoding.  In practice, most HTTP header
   field values use only a subset of the US-ASCII charset [USASCII].
   Newly defined header fields SHOULD limit their field values to
   US-ASCII octets.  A recipient SHOULD treat other octets in field
   content (obs-text) as opaque data.
```

US-ASCII FTW! Thanks :)

Looking forward for the patch.

**Comment by twisted-trac** [comment]
|[<img alt="pawelmhm's avatar" src="https://avatars.githubusercontent.com/u/2700942?s=50" width="50" height="50">](https://github.com/pawelmhm)<a name="note_8"></a>|@pawelmhm commented|
|-|-|

ok no problem! 

Attached patch for issue.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

(In [46282]) Branching to http-server-invalid-method-8102-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

Changes applied and sent to buildbot... waiting for test to be done and will merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46295]) Merge http-server-invalid-method-8102-2: Fix unhandled error for invalid HTTP method name.

Author: pawelmhm
Reviewer: adiroiban
Fixes: [#8102](https://github.com/twisted/twisted/issues/8102)
